documentation/internal/Development-Workflow.md

```diff
@@ -0,0 +1,311 @@
+
+**Important**: No need to do anything different to query from the cache instead of the server. Just a simple query will do it. Mutations on the other hand require some addition - after sending data to the server we have to update the cache.
+
+_How Apollo keeps the cache synced with the server?_
+
+Apollo client provides this cache management module, which will store all the data query or mutation results after it is fetched from the server. It requires you add the `id` of each entity - and the related entities included in the fetched data - or another identifier that you have specified before to the cache management instance if not using the `id`. The caches stores one instance of each entity queried (refreshing would clear the cache). And for each relationship it keeps a reference to its entity. In the end the remote states has one instance of each entity already fetched from the server.
+
+_Why do we update the cache after mutations?_
+
+When we create a new entity that has a relationshio with some existing in cache we need to also update the cache with the added entity! This is to avoid having to create a `refetch` to get this back from the server and keep the local cache up-to-date. The same way, after editing an entity which is already in cache we need to remember to update this entity in the cache.
+
+##### Example of remote data
+
+- Current Application
+- Application Sections (related enetities - Application)
+  - Template Sections (related enetities - Application Section)
+- Application Responses (related entities - Application)
+  - Template element (related entities - Application Response)
+
+### Local states
+
+We are using 2 options for local states:
+
+- local variables (with `useState`) passed as `props` to other components
+- context areas (with `useContext`) for sharing between components
+
+##### Example of local state
+
+**TODO**: Rename ApplicationPage => ApplicationPageWrapper (if we decide that we want to keep the wrapper thingy).
+
+**To be created** => In ApplicationPageWrapper:
+
+- List of all questions and responses
+
+How to use:
+
+1. `ApplicationPage` calls initial query to load the whole application, section and responses.
+2. `ApplicationPage` calls a custom hook to create the list of questions and responses based on query.
+3. Custom hook returns the result object Q&R with all the question elements with reduced fields (after running the `evaluateExpression` on each question) and the responses by the user.
+
+The object returned has for each Q&R (Question and Response):
+
+- Question code (key)
+- Type: Information | Question - if information none of the next fields are filled!
+- isRequired: boolean
+- isVisible: boolean
+- isValid: boolean
+- ResponseId: number
+- ResponseJSON: JSON
+- ResponseValue: string
```

Comment:
After doing [this](https://github.com/openmsupply/application-manager-server/pull/95) in response to [this](https://github.com/openmsupply/application-manager-server/issues/61), and working on this local state Q & R object, I think the full list of fields will be:

- Question code (key)
- Category: Information | Question
- isRequired: boolean
- isVisible: boolean
- isValid: boolean
- isEditable: boolean
- validationMessage: string
- ResponseId: number
- ResponseJSON: JSON
- ResponseValue: string

Also, regarding your comment "If information none of the next fields are filled" -- `isVisible` can still be relevant!

Reply:
I have updated this, also talking about 2 objects that should be returned by this custom hook.